### DISCUSSION TITLE
NvmMangement in the LoraWan Stack

### DISCUSSION THREAD

**Discussion by Adamzat** [Lora-net/LoRaMac-node#1120] [Q&A] [answered]
Hi all,

I have a question about the NVM context management.

Actually, we want to optimize the access to the EEPROM of our devices. Indeed, we would like to know:
- If we must store all the data managed by the LoraWan Stack (described in LoRaMacNvmData_t)?
For example, it seems that the data used for the Join are necessary (like the DevNonce). But I am not sure for the others?

- What is the frequency that the data is updated by the LoraWanStack. Is it every uplink or reset? or it is random?

Thank you.
Best regards.

**Comment by bwesen** [comment, marked as answer]
In my recent experience it seems like it tries to write hundreds of EEPROM bytes for every packet received (which takes seconds on the STM32L0). I had to disable context management completely...  there are not many comments anywhere in the stack so I couldn't dig deeper.

**Reply by Adamzat** [reply]
Yes, it writes about 1452 bytes in the EEPROM the first time which is huge. However, I don't think that we can disable it completely, because some certification tests will not pass (to be compliant with spec. 1.0.4).  
The Devnonce for example should be stored and incremented after reset.

**Reply by kjendal** [reply]
For an ABP device, all state context must be stored to recover from a reset.  This includes Frame Counters and RX Parameters.
An OTAA device only needs to store this information if it wont re-join after the reset.  The OTAA device also needs to store the DevNonce and JoinNonce used during the Join Procedure.  The specifics are documented in TR007 on the LoRa Alliance hub.

Impact to the EEPROM can be mitigated at the cost of some gaps in the counters (Frame Counters and DevNonce).  For example, if after the reset you advance the stored Frame Counter by 10 (instead of 1), you only need to write to the EEPROM every 10th ukplink.  But you could do this every 100th, 1000th, etc.

**Reply by bwesen** [reply]
I'm using OTAA and it still stores this every uplink. But maybe you mean that the stack could be improved to not do that?

I guess I'm safe then as we do a Join on every boot, if I never store anything :)

**Reply by kjendal** [reply]
You only need to store the DevNonce (and ideally JoinNonce) used during the Join Procedure.

**Reply by bwesen** [reply]
But it works fine with all storing disabled (and worked fine for several years using the older stack). What has changed now that makes it necessary?

**Reply by Adamzat** [reply]
Thank you Dave,
It is clear, OK for DevNonce and the JoinNonce.
However, I did not understand if we should also save the Frame counter as you describe above in OTAA mode?

**Reply by kjendal** [reply]
Frame counter is not required if you will join after reboot

**Reply by kjendal** [reply]
To improve the security posture, in LW 1.0.4 DevNonce is required to be monotonic (always increasing).  This allows the JS to systematically guarantee there has been no replay attack.  JoinNonce is required to be monotonic in LW 1.1+ for the same reason towards the end-device.  The support this requirement, the end-device must store these values.

**Reply by bwesen** [reply]
Yeah I think I understand the reasoning, but I fail to see how it can work in practice. No device can keep this stored correctly forever and no Lora GW can reliably store it forever either. If either of them loses it, should I ask my customers to throw away their products? :)

I wonder if there is a way to tell the Multitech Conduit GW to treat all devices as 1.0.3 and simply ignore this mess.. and vice versa for the device stack..

**Reply by kjendal** [reply]
I dont understand your point.  The end-device and the Join Server need to store a single value such that they do not use a previous nonce (as they are a counter).  That is not hard to do, in fact it is easier than what they were trying to do previously, which is to keep a list of previously used values.  The gateway has nothing to do with it.

**Reply by bwesen** [reply]
Well the device can be factory reset or reflashed, the EEPROM CRC can break, miswrites or misreads happen. The gateway can get reflashed (the "Join Server" is just the gateway in the Multitech case). If this really was enforced, nobody would sell a product that can get bricked and not reset.. 

I'm sure there must be a way to just disable this enforcement in both the GW and the device.
